### ISSUE TITLE
Bring back typedef property expansion

### ISSUE THREAD

**Issue by mjbvz** [Microsoft/TypeScript#25784]
_From @zavr-1 on July 18, 2018 20:25_

In the newer version, I can only see the type of an argument which is an object, compared to how it was expanded in the previous versions.

![test3](https://user-images.githubusercontent.com/21156791/42904849-3f94249c-8ace-11e8-91bb-8a09063e9d30.png)

It was actually very useful to see all properties on hover which served as a documentation. 

![test4](https://user-images.githubusercontent.com/21156791/42905747-d387673e-8ad0-11e8-92c8-ddf31e9d7f92.png)

I think if you hide the details of the type, you should also provide means to expand its properties via a UI, or provide a setting for it. 

_Copied from original issue: Microsoft/vscode#54609_

**Comment by mjbvz** [comment]
@zavr-1  do you see the @param properties if you scroll the hover?

**Comment by mjbvz** [comment]
Yes, I think I misunderstood this specific issue but having a way to expand types dynamically in hover would be very helpful. We have no concept of this in VS Code currently. Visual Studio might though

**Comment by mhegazy** [comment]
VS does not either. i think eclipse does.

**Comment by zavr-1** [comment]
@mhegazy @mjbvz 
I can achieve this expansion by writing 

```js
/** @type {DoneStatus} */
var ns = {
}

/**
 * @typedef {{ status: Status,  m(n: number): void }} DoneStatus
 * @typedef {{
  hello: 'world'
}} Status */
```

![test](https://user-images.githubusercontent.com/21156791/43013714-b458723c-8c41-11e8-81e3-6e711a2be38d.png)

However, there is no way to provide description of a property for auto-complete hints 

![test](https://user-images.githubusercontent.com/21156791/43013786-eca4abe2-8c41-11e8-8954-ff5db8248f1c.png)

Therefore, the quickest solution _before_ providing expansion of types is to allow writing `JSDoc` where a type is specified both with `{{ ... }}` to allow for expanded properties in preview, and has `@prop` for description of properties in the auto-complete. 

```js
/**
 * @typedef {{ status: Status,  m(n: number): void }} DoneStatus
 * @prop {Status} status The status
 * @typedef {{
  hello: 'world'
}} Status */
```

(Note the additional `@prop` in the typedef).

**Comment by zavr-1** [comment]
@weswigham I don't think that your change from [here](https://github.com/Microsoft/TypeScript/commit/5bf6e30f8e839d8e8887ec4ffaa63e70c02e29a4) which is the point of this topic will be reverted, but is it possible to at least implement the suggestion in the post above ^? thanks.

**Comment by weswigham** [comment]
@zavr-1 That PR brought JSDoc into the same style of printout that we used for TS, but the core issue is that when a type is aliased within quickinfo, there's no way to inspect what's "inside the alias" right at that point, as it were. Hence @mhegazy and @mjbvz 's remarks on new UI to enable that in vscode/VS. We've talked about it in the past - it seems generally useful for inspecting large type hierarchies.

**Comment by zavr-1** [comment]
@weswigham OK but why is it possible to inspect what's inside using 
```
@typedef {{ prop: TypeB }} TypeA
```
but not 
```
@typedef {Object} TypeA
@prop {TybeB} prop
```
What I'm saying OK fine I get the point of aliases for large type hierarchies, but could you not allow me to write `{{ prop: TypeB }}` **AND** define a property under `@typedef`?

**Comment by AoDev** [comment]
Hi, I created a similar issue that was closed for duplicate. (#41096 ) 
Here the last comment is from 2 years ago and there is another related task (#38040)

I'd like a quick update on where we are with this.

**Comment by rgembalik** [comment]
I'd love to see this happen as internal typedefs are usually quite important in terms of their structure. A quick set of ideas to please both people who need it and people who prefer the simplified view:

1. Have an option in the settings to select behavior
2. Have a link-like button in the hint popup "Expand types"/"Collapse types"
3. Allow clicking on the type name to expand it or bring its own pop-up
4. All of the above?

**Comment by brillout** [comment]
Related feature request: https://github.com/microsoft/TypeScript/issues/56010.

**Comment by Mng12345** [comment]
Is there any progress?

**Comment by aiday-mar** [comment]
In relation to this, recently we have merged into VS Code a proposed API for expanding and contracting hovers (see PR here https://github.com/microsoft/vscode/pull/210472), so if the TypeScript server adopts quick info expansion, we could implement this in VS Code too.

**Comment by 0xdevalias** [comment]
> so if the TypeScript server adopts quick info expansion, we could implement this in VS Code too

See also:

- https://github.com/microsoft/TypeScript/issues/59029
- https://github.com/microsoft/TypeScript/pull/61492

**Comment by skorphil** [comment]
Version 1.104. Expandable view doesn't seem to work.
Any updates? 

BTW, I see this feature working in github.dev
